src/mito2/src/sst/index/creator.rs

```diff
@@ -0,0 +1,265 @@
+use crate::sst::index::creator::temp_provider::TempFileProvider;
+use crate::sst::index::store::InstrumentedStore;
+use crate::sst::index::{
+    INDEX_BLOB_TYPE, MIN_MEMORY_USAGE_THRESHOLD, PIPE_BUFFER_SIZE_FOR_SENDING_BLOB,
+};
+use crate::sst::location::{self, IntermediateLocation};
+
+type ByteCount = usize;
+type RowCount = usize;
+
+/// Creates SST index.
+pub struct SstIndexCreator {
+    /// Directory of the region.
+    region_dir: String,
+    /// ID of the SST file.
+    sst_file_id: FileId,
+
+    /// The store to write intermediate files and index files.
+    store: InstrumentedStore,
+    /// The index creator.
+    index_creator: Box<dyn InvertedIndexCreator>,
+    /// The provider of intermediate files.
+    temp_file_provider: Arc<TempFileProvider>,
+
+    /// Codec for decoding primary keys.
+    codec: IndexValuesCodec,
+    /// Reusable buffer for encoding index values.
+    value_buf: Vec<u8>,
+
+    /// Statistics of index creation.
+    stats: Statistics,
+    /// Whether the index creation is aborted.
+    aborted: bool,
+}
+
+impl SstIndexCreator {
+    /// Creates a new `SstIndexCreator`.
+    /// Should ensure that the number of tag columns is greater than 0.
+    pub fn new(
+        region_dir: String,
+        sst_file_id: FileId,
+        metadata: &RegionMetadataRef,
+        object_store: ObjectStore,
+        memory_usage_threshold: Option<usize>,
+        row_group_size: NonZeroUsize,
+    ) -> Self {
+        let store = InstrumentedStore::new(object_store);
+
+        // `memory_usage_threshold` is the total memory usage threshold of the index creation,
+        // so we need to divide it by the number of columns
+        let memory_threshold = memory_usage_threshold.map(|threshold| {
+            (threshold / metadata.primary_key.len()).max(MIN_MEMORY_USAGE_THRESHOLD)
+        });
+        let temp_file_provider = Arc::new(TempFileProvider::new(
+            IntermediateLocation::new(&region_dir, &sst_file_id),
+            store.clone(),
+        ));
+        let sorter = ExternalSorter::factory(temp_file_provider.clone() as _, memory_threshold);
+        let index_creator = Box::new(SortIndexCreator::new(sorter, row_group_size));
+
+        let codec = IndexValuesCodec::from_tag_columns(metadata.primary_key_columns());
+        Self {
+            region_dir,
+            sst_file_id,
+            store,
+            codec,
+            index_creator,
+            temp_file_provider,
+
+            value_buf: vec![],
+
+            stats: Statistics::default(),
+            aborted: false,
+        }
+    }
+
+    /// Updates index with a batch of rows.
+    /// Garbage will be cleaned up if failed to update.
+    pub async fn update(&mut self, batch: &Batch) -> Result<()> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        if batch.is_empty() {
+            return Ok(());
+        }
+
+        if let Err(update_err) = self.do_update(batch).await {
```

Comment:
If the program crashes after line 130, but before line 132 where the `do_cleanup` occurs, what will happen to external garbage? Will it be cleaned up the next time?

Reply:
According to my implementation, the garbage intermediate files caused by abnormal exit will be left and will not be cleaned up next time.

Rather than doing some additional processing in the cleanup logic, it is better to clean up the intermediate directory as a whole during initialization. However, this cleanup logic is not in the scope of this PR.

Reply:
Ok 👍🏻